### PR TITLE
Limit maximum pyOpenSSL version

### LINKED ISSUE #1373: Limit pyOpenSSL version to 17.5.0
The [pyOpenSSL](https://pypi.org/project/pyOpenSSL/) package currently [required](https://github.com/RedHatInsights/insights-core/blob/95d79585f06a99dc6c6e910bad0b2d896a9b9d79/setup.py#L42) does not support Python 2.6 in its most recent version [18.0.0](https://pypi.org/project/pyOpenSSL/18.0.0/). However only a one step previous version [17.5.0](https://pypi.org/project/pyOpenSSL/17.5.0/) still supports it and installs just fine.

If the tests don’t fail and there are no other issues (security, since it’s SSL?), we should note this specific version in the [dependency list](https://github.com/RedHatInsights/insights-core/blob/95d79585f06a99dc6c6e910bad0b2d896a9b9d79/setup.py#L42). Otherwise the package is not installable on systems with this legacy version of Python (RHEL6).

### PR DESCRIPTION
**Pull request by Glutexo**
The [most recent version](
https://pypi.org/project/pyOpenSSL/18.0.0/) of [pyOpenSSL](https://pypi.org/project/pyOpenSSL/) fails to install on Python 2.6. Its cryptography dependency cannot be satisfied. Forced installation of [the last version](https://pypi.org/project/pyOpenSSL/17.5.0/) that still supports 2.6.

To use the downgraded [pyOpenSSL](
https://pypi.org/project/pyOpenSSL/) in Jenkins pipeline I updated [Jenkinsfile](https://github.com/RedHatInsights/insights-core/pull/1386/files#diff-58231b16fdee45a03a4ee3cf94a9f2c3) to use a virtual environment just like other pipeline branches do. This is necessary to do at once with the downgrade, because the global environment in the image prevents proper installation of [pyOpenSSL](https://pypi.org/project/pyOpenSSL/).

~~This PR should help #1369.~~ (Merged.) Fixes #1373.

No _pytest_ tests fail after the downgrade neither on Python 2.6, 2.7 nor 3.

### CODE REVIEW

**Comment by Glutexo** [comment]
Thanks to the fact that #1369 got merged, this works and is mergeable too.

**Comment by Glutexo** [comment]
It looks like @gravitypriest and @jneedle may have some new opinions on this.

**Comment by jneedle** [comment]
Yeah, see https://bugzilla.redhat.com/show_bug.cgi?id=1615598 and https://access.redhat.com/solutions/3563671 and tread carefully.

**Comment by Glutexo** [comment]
From what I understood from the console dumps in the BZ comments, the _pyOpenSSL_ installed by _yum_ is version _17.3.0_. That means that my change to required _pyOpenSSL_ [version](https://github.com/RedHatInsights/insights-core/pull/1386/files#diff-2eeaed663bd0d25b7e608891384b7298R42) would make some dependency issues: either breaking the install, or forcing either upgrade or side-by-side install of 17.5.0.

I downgraded the _pyOpenSSL_ dependency for compatibility with Python 2.6 when installing in a virtual environment. It doesn’t work with the most recent version that _setuptools_/_pip_ install when that package is not detected. If _insights-client_ works with that older (_17.3.0_) version of _pyOpenSSL_ then this can be resolved by amending the downgrade not to force this specific version, but to cap it to the last working one. Instead of `'pyOpenSSL==17.5.0'` there can be `'pyOpenSSL<=17.5.0'`.

You (@jneedle) most probably know much more about RHEL packaging. Am I right that `'pyOpenSSL==17.5.0'` would cause problems, but `'pyOpenSSL<=17.5.0'` would be OK?

**Comment by Glutexo** [comment]
I rebased this on the current [_master_](https://github.com/RedHatInsights/insights-core/commit/4f07378d044c63559fadd709fbe7809ee5fe667c) and hopefully fixed it.

It’s possible (and probable) that _pyOpenSSL_ is installed on the system via RPM instead of PIP. On RHEL 6 that means version _0.13.1_. Because of that I changed the dependency [string](https://github.com/RedHatInsights/insights-core/pull/1386/files#diff-2eeaed663bd0d25b7e608891384b7298R43) from `==17.5.0` to `<=17.5.0`. Removing the lower limit makes PIP use the already present version as it did before the change. The upper limit makes it work in a virtual environment.

I had to include [_pycparser_] version [limit](https://github.com/RedHatInsights/insights-core/pull/1386/files#diff-2eeaed663bd0d25b7e608891384b7298R42). Just a few days ago, a new [version](
https://pypi.org/project/pycparser/2.19)/ appeared that does not install on Python 2.6. Without limiting, it‘s not possible to install [_pyOpenSSL_](https://pypi.org/search/?q=pyOpenSSL).

**Comment by Glutexo** [comment]
Jenkins is [failing](http://10.8.252.209:8080/blue/organizations/jenkins/CI-PR-insights-core/detail/CI-PR-insights-core/563/pipeline/) on  _insights-tests/bootstrap.sh_. May I ask @eduardocerqueira to look at this, please?

Also asking @jneedle or @kylape to review whether it’s possible to merge this simple change without breaking anything.

**Comment by esakaiev** [comment]
test this please

**Comment by esakaiev** [comment]
Some issues with our openstack, Jenkins can't reach any of the new server instances. 
Asked @jneedle for help with that. It could be the reason for the failure of every our pipeline.

**Comment by Glutexo** [comment]
@esakaiev Thanks for looking into this! Please trigger the tests here once it’s fixed and let me know then.

**Comment by esakaiev** [comment]
test this please

**Comment by esakaiev** [comment]
It seems, that I can't restart Jenkins pipeline from here, due to "The base branch restricts merging to authorized users.". @Glutexo could you trigger it again, please?

**Comment by Glutexo** [comment]
Rebased on [current _master_](https://github.com/RedHatInsights/insights-core/commit/7f1a9304ad649a15f9380699b00df1a19e311ead). The downgraded versions were already used in the RHEL 6 [pipeline](https://github.com/RedHatInsights/insights-core/blob/7f1a9304ad649a15f9380699b00df1a19e311ead/Jenkinsfile#L13). When downgrading right in the [_setup.py_](
https://github.com/RedHatInsights/insights-core/pull/1386/files#diff-2eeaed663bd0d25b7e608891384b7298), the hack in [_Jenkinsfile_](
https://github.com/RedHatInsights/insights-core/pull/1386/files#diff-58231b16fdee45a03a4ee3cf94a9f2c3) is no longer necessary. So removed.

May I ask @eduardocerqueira or @bfahr to take a look?

**Comment by Glutexo** [comment]
Closed by accident.

**Comment by kylape** [comment]
@Glutexo Can you summarize what problem you are trying to solve with this PR?  I know we have some RPM confusion, but this seems specific to dev environments.

**Comment by Glutexo** [comment]
More packages need downgrading to work with Python 2.6:

• [_pyOpenSSL_](https://pypi.org/project/pyOpenSSL/) requires [_idna_](https://pypi.org/project/idna/) through [_cryptography_](https://pypi.org/project/cryptography/). The last version that supports Python 2.6 is [2.7](https://pypi.org/project/idna/2.7/). `idea<=2.7` needs to be added to the [_client_](https://github.com/RedHatInsights/insights-core/blob/ac47281f9a83bb762982b14fb3b0824eb70d0c04/setup.py#L48) bundle.
• The last version of [_ansible_](https://pypi.org/project/ansible/) that supports Python 2.6 is [2.6.13](https://pypi.org/project/ansible/2.6.13/). Its line must in the [_cluster_](https://github.com/RedHatInsights/insights-core/blob/ac47281f9a83bb762982b14fb3b0824eb70d0c04/setup.py#L74) bundle must be amended to `ansible<=2.6.13`.
• [_pandas_](https://pypi.org/project/pandas/) supports Python 2.6 only until [_0.17.1_](https://pypi.org/project/pandas/0.17.1/). Its line in the [_cluster_](https://github.com/RedHatInsights/insights-core/blob/ac47281f9a83bb762982b14fb3b0824eb70d0c04/setup.py#L74) bundle must be amended to `pandas<=0.17.1`.
• [_numpy_](https://pypi.org/project/numpy/) required by [_pandas_](https://pypi.org/project/pandas/) supports Python 2.6 only until [1.11.3](https://pypi.org/project/numpy/1.11.3/). For some reason, adding `numpy<=1.11.3` to the [_cluster_](https://github.com/RedHatInsights/insights-core/blob/ac47281f9a83bb762982b14fb3b0824eb70d0c04/setup.py#L74) bundle doesn’t solve the problem. But it’s ok when installed manually by `pip install 'numpy<=1.11.3'`. I didn’t investigate any further.
• The last version of [_colorama_](https://pypi.org/project/colorama/) that supports Python 2.6 is [0.3.9](https://pypi.org/project/colorama/0.3.9/). Its line in the [_cluster_](https://github.com/RedHatInsights/insights-core/blob/ac47281f9a83bb762982b14fb3b0824eb70d0c04/setup.py#L74) bundle must be amended to `colorama<=0.3.9`.

For some reason I had to manually install [_Cython_](https://pypi.org/project/Cython/) by running `pip install Cython`. It is required by [_pandas_](https://pypi.org/project/pandas/), but it didn’t install it automatically. Furthermore, I had to do `yum install gcc-c++` to install [_Cython_](https://pypi.org/project/Cython/) with its native extensions.

After doing all that and removing [_docs_](https://github.com/RedHatInsights/insights-core/blob/ac47281f9a83bb762982b14fb3b0824eb70d0c04/setup.py#L58) I was able to install the [_develop_](https://github.com/RedHatInsights/insights-core/blob/ac47281f9a83bb762982b14fb3b0824eb70d0c04/setup.py#L53) bundle on RHEL6. I’m not going to invest any more time into this now. But either me or anybody can hopefully use this information in the future. I neither verified whether the described downgrades break anything, nor I ran the test suite!

**Comment by jhjaggars** [comment]
A more recent PR removes this dependency.

**Comment by Glutexo** [comment]
Just linking #2099 for reference.
